tensorflow/core/kernels/dml_softmax_ops.cc

```diff
@@ -0,0 +1,109 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/common_runtime/dml/dml_operator_helper.h"
+#include "tensorflow/core/common_runtime/dml/dml_util.h"
+#include "tensorflow/core/framework/register_types.h"
+#include "tensorflow/core/kernels/dml_kernel_wrapper.h"
+#include "tensorflow/core/kernels/dml_ops_common.h"
+
+namespace tensorflow {
+
+class DmlSoftmaxXentWithLogitsKernel : public DmlKernel {
+ public:
+  using InitHelper = NoOpInitializationHelper;
+  explicit DmlSoftmaxXentWithLogitsKernel(
+      DmlKernelConstruction* ctx, const NoOpInitializationHelper* init_helper) {
+    CHECK(ctx->GetInputCount() == 2);
+    CHECK(ctx->GetOutputCount() == 2);
+
+    DmlKernelParams params;
+    params.kernel_input_indices = {0, 1};
+    params.kernel_output_indices = {0, 1};
+
+    DmlKernelTensors tensors = GetTensorInfos(ctx, params);
+    auto input_descs = GetDmlTensorDescs(tensors.inputs);
+
+    auto scope = dml::Scope(ctx->GetDmlDevice());
+
+    // logits: batch_size, num_classes.
+    // labels: batch_size, num_classes.
+    auto logits = dml::InputTensor(scope, 0, input_descs[0]);
+    auto labels = dml::InputTensor(scope, 1, input_descs[1]);
+
+    auto input_sizes = labels.GetOutputDesc().sizes;
+
+    // The strides we need to set to broadcast class across an entire tensor
+    dml::TensorDesc::Dimensions broadcast_c_strides = {/*NoSense*/ 0,
+                                                       /*NoSense*/ 0,
+                                                       /*kBatchDim*/ 1,
+                                                       /*kClassDim*/ 0};
+
+    // The strides we need to set to broadcast class across an entire tensor
+    dml::TensorDesc::Dimensions reduce_strides = {0, 1, 3};
+
+    // max_logits along classes.
+    auto logits_max =
+        dml::Reduce(logits, DML_REDUCE_FUNCTION_MAX, reduce_strides);
+    auto logits_max_bcast =
+        dml::Reinterpret(logits_max, input_sizes, broadcast_c_strides);
+
+    // logits - max_logits.
+    auto shifted_logits = logits - logits_max_bcast;
+
+    // exp(logits - max_logits)
+    auto exp_shifted_logits = dml::Exp(shifted_logits);
+
+    // sum(exp(logits - max_logits)) along classes.
+    auto sum_exp = dml::Reduce(exp_shifted_logits, DML_REDUCE_FUNCTION_SUM,
+                               reduce_strides);
+
+    // log(sum(exp(logits - max_logits)))
+    auto log_sum_exp = dml::Log(sum_exp);
+    auto log_sum_exp_bcast =
+        dml::Reinterpret(log_sum_exp, input_sizes, broadcast_c_strides);
+
+    // sum(-labels *
+    //    ((logits - max_logits) - log(sum(exp(logits - max_logits)))))
+    auto sub = shifted_logits - log_sum_exp_bcast;
+    auto mul = -1.0f * labels * sub;

```

Comment:
you can simply do -labels here

Reply:
Actually, scratch that. You can remove an operation by doing `sub = log_sum_exp_bcast - shifted_logits` instead. After all, `-1 * (x - y) = y - x`